tsdb/head_test.go

```diff
@@ -4744,3 +4744,61 @@ func TestGaugeFloatHistogramWALAndChunkHeader(t *testing.T) {
 
 	checkHeaders()
 }
+
+func TestSnapshotAheadOfWALError(t *testing.T) {
+	head, _ := newTestHead(t, 120*4, false, false)
+	entries := []interface{}{
+		[]record.RefSeries{
+			{Ref: 10, Labels: labels.FromStrings("a", "1")},
+			{Ref: 11, Labels: labels.FromStrings("a", "2")},
+			{Ref: 100, Labels: labels.FromStrings("a", "3")},
+		},
+		[]record.RefSample{
+			{Ref: 0, T: 99, V: 1},
+			{Ref: 10, T: 100, V: 2},
+			{Ref: 100, T: 100, V: 3},
+		},
+		[]record.RefSeries{
+			{Ref: 50, Labels: labels.FromStrings("a", "4")},
+			// This series has two refs pointing to it.
+			{Ref: 101, Labels: labels.FromStrings("a", "3")},
+		},
+		[]record.RefSample{
+			{Ref: 10, T: 101, V: 5},
+			{Ref: 50, T: 101, V: 6},
+			{Ref: 101, T: 101, V: 7},
+		},
+		[]tombstones.Stone{
+			{Ref: 0, Intervals: []tombstones.Interval{{Mint: 99, Maxt: 101}}},
+		},
+		[]record.RefExemplar{
+			{Ref: 10, T: 100, V: 1, Labels: labels.FromStrings("traceID", "asdf")},
+		},
+	}
+
+	w, _ := wlog.NewSize(nil, nil, head.wal.Dir(), 32768, false)
+	populateTestWAL(t, w, entries)
+
+	head.opts.EnableMemorySnapshotOnShutdown = true
+	require.NoError(t, head.Close()) // This will create a snapshot.
+	require.NoError(t, w.Close())
+
+	snapDir, _, _, err := LastChunkSnapshot(head.opts.ChunkDirRoot)
+	require.NoError(t, err)
+
+	// Corrupt the filename by increasing the snapshot index beyond the WAL index
```

Comment:
It is not a corruption. And renaming the file wont test the actual scenario. To test the actual scenario, we should
* add data, while creating a few WAL files
* then take a snapshot when shutting down
* then clear WAL, and start TSDB without snapshot enabled, so that it doesn't read from it. Then add more data, which will create WAL file from 0, i.e. before snapshot.
* Now restart with snapshots enabled. And verify that only the samples added in the second round were present, hence skipping the snapshot.